LotteMusicFriend test scenario fails: "attempting to recurse into an non-table/document" when dumping

```text
majo@cspmilkmini ~/W/tomlantic (main)> nix flake check
error: builder for '/nix/store/glgrzhwsy64qfqy051akla10kqjkg2rw-tomlantic-check-tests.drv' failed with exit code 1;
       last 25 log lines:
       >  ... 314 -> (unresolved)
       >  ... local -> `/nix/store/v6chcvgf38mhp8pjqn4gcrdi5s8hjxmp-tomlantic-testing-env/bin/python3.13 ... `
       > tomlantic tests: running showcase.py@pylocal...
       > tomlantic tests: running test-grassplain.py@pylocal...
       > tomlantic tests: running test-lottemusicfriend.py@pylocal...
       > - test-lottemusicfriend.py@pylocal
       >   ... Traceback (most recent call last):
       >   ...   File "/nix/store/8hbr47r071qgp8wnm5mhhdxdi2kxd1m2-y8grddpkrhdpl4399qc2yqcaf905wl10-source/tests/test-lottemusicfriend.py", line 109, in <module>
       >   ...     test()
       >   ...     ~~~~^^
       >   ...   File "/nix/store/8hbr47r071qgp8wnm5mhhdxdi2kxd1m2-y8grddpkrhdpl4399qc2yqcaf905wl10-source/tests/test-lottemusicfriend.py", line 105, in test
       >   ...     print(dumps(model.model_dump_toml()))
       >   ...                 ~~~~~~~~~~~~~~~~~~~~~^^
       >   ...   File "/nix/store/v6chcvgf38mhp8pjqn4gcrdi5s8hjxmp-tomlantic-testing-env/lib/python3.13/site-packages/tomlantic/tomlantic.py", line 528, in model_dump_toml
       >   ...     apply_model_differences(self.__original_model, self.model, document)
       >   ...     ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >   ...   File "/nix/store/v6chcvgf38mhp8pjqn4gcrdi5s8hjxmp-tomlantic-testing-env/lib/python3.13/site-packages/tomlantic/tomlantic.py", line 522, in apply_model_differences
       >   ...     apply_model_differences(og_val, cu_val, toml_cu_key)
       >   ...     ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >   ...   File "/nix/store/v6chcvgf38mhp8pjqn4gcrdi5s8hjxmp-tomlantic-testing-env/lib/python3.13/site-packages/tomlantic/tomlantic.py", line 518, in apply_model_differences
       >   ...     assert isinstance(toml_cu_key, (TOMLDocument, tomlitems.Table)), (
       >   ...            ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >   ... AssertionError: key service: attempting to recurse into an non-table/document
       >
       > tomlantic tests: failed 1/3 (33.33%) tests
       For full logs, run:
         nix log /nix/store/glgrzhwsy64qfqy051akla10kqjkg2rw-tomlantic-check-tests.drv
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -19,6 +19,7 @@ license-files = ["LICENCE*"]
 
 [dependency-groups]
 dev = [
+    "basedpyright>=1.37.1",
     "isort>=6.0.1",
     "mypy>=1.18.1",
     "ruff>=0.13.0",

--- a/tests/test-array-of-tables.py
+++ b/tests/test-array-of-tables.py
@@ -1,0 +1,77 @@
+# /// script
+# requires-python = ">=3.9"
+# dependencies = [
+#     "pydantic",
+#     "tomlantic @ ${PROJECT_ROOT}/",
+#     "tomlkit",
+# ]
+# ///
+
+
+from pydantic import BaseModel
+from tomlkit import dumps, loads  # pyright: ignore[reportUnknownVariableType]
+
+import tomlantic
+
+
+class Item(BaseModel):
+    name: str
+    value: int
+    active: bool
+
+
+class ConfigWithArray(BaseModel):
+    title: str
+    items: list[Item] = []
+
+
+ARRAY_TEST_EXAMPLE = """
+title = "Test Array"
+
+[[items]]
+name = "first"
+value = 1
+active = true
+
+[[items]]
+name = "second"
+value = 2
+active = false
+"""
+
+
+def test():
+    """test array of tables (list[BaseModel])"""
+
+    toml = tomlantic.ModelBoundTOML(ConfigWithArray, loads(ARRAY_TEST_EXAMPLE))
+
+    # Test 1: initial dump should work
+    _ = dumps(toml.model_dump_toml())
+
+    # Test 2: modify existing item
+    toml.model.items[0].value = 10
+    dumped = dumps(toml.model_dump_toml())
+    assert "value = 10" in dumped
+    assert "value = 2" in dumped
+
+    # Test 3: add new item
+    toml.model.items.append(
+        Item(
+            name="third",
+            value=3,
+            active=True,
+        )
+    )
+    dumped = dumps(toml.model_dump_toml())
+    assert 'name = "third"' in dumped
+
+    # Test 4: remove item (test array rebuild)
+    toml.model.items.pop(0)
+    dumped = dumps(toml.model_dump_toml())
+    assert 'name = "first"' not in dumped
+    assert 'name = "second"' in dumped
+    assert 'name = "third"' in dumped
+
+
+if __name__ == "__main__":
+    test()

--- a/tests/test.py
+++ b/tests/test.py
@@ -27,7 +27,9 @@ def resolve_binaries() -> dict[str, list[str]]:
             getenv("name", default="").startswith("nix"),
             getenv("NIX_STORE", default=None) is not None,
             getenv("IN_NIX_SHELL", default=None) is not None,
-            str(which("python")).startswith(getenv("NIX_STORE", default="")),
+            str(which("python")).startswith(
+                getenv("NIX_STORE", default="zsxgdfchgvjhbj")
+            ),
         ]
     ):
         # not in nix, detect uv or die

--- a/tomlantic/tomlantic.py
+++ b/tomlantic/tomlantic.py
@@ -22,8 +22,11 @@ from typing import (
 from pydantic import BaseModel, ValidationError
 from pydantic_core import ErrorDetails
 from tomlkit import TOMLDocument
+from tomlkit import array as toml_array
 from tomlkit import items as toml_items
 from tomlkit import table as toml_table
+from tomlkit.container import OutOfOrderTableProxy
+from tomlkit.items import AoT as toml_items_AoT
 
 if TYPE_CHECKING:
     from typing_extensions import override  # pyright: ignore[reportMissingModuleSource]
@@ -573,6 +576,44 @@ def _get_model_field(
         return default
 
 
+def _is_dict_of_basemodels(value: object) -> bool:
+    if not isinstance(value, dict):
+        return False
+    if not value:
+        return True
+    return all(isinstance(v, BaseModel) for v in value.values())
+
+
+def _is_list_of_basemodels(value: object) -> bool:
+    if not isinstance(value, list):
+        return False
+    if not value:
+        return True
+    return all(isinstance(item, BaseModel) for item in value)
+
+
+def _dicts_of_basemodels_differ(
+    original: dict[str, BaseModel], current: dict[str, BaseModel]
+) -> bool:
+    if original.keys() != current.keys():
+        return True
+    for key in original:
+        if original[key] != current[key]:
+            return True
+    return False
+
+
+def _lists_of_basemodels_differ(
+    original: list[BaseModel], current: list[BaseModel]
+) -> bool:
+    if len(original) != len(current):
+        return True
+    for i, (orig_item, curr_item) in enumerate(zip(original, current)):
+        if orig_item != curr_item:
+            return True
+    return False
+
+
 class ModelBoundTOML(Generic[M]):
     """
     glue class for pydantic models and tomlkit documents
@@ -659,6 +700,30 @@ class ModelBoundTOML(Generic[M]):
         """
         document = deepcopy(self.__document)
 
+        def _apply_basemodel_differences(
+            original_model: BaseModel,
+            current_model: BaseModel,
+            toml: Union[TOMLDocument, toml_items.Table],
+        ) -> None:
+            for original_kv, current_kv in zip(original_model, current_model):
+                og_key, og_val = original_kv  # pyright: ignore[reportAny] (from pydantic)
+                cu_key, cu_val = current_kv  # pyright: ignore[reportAny] (from pydantic)
+
+                if isinstance(cu_val, BaseModel):
+                    if cu_key not in toml:
+                        toml[cu_key] = toml_table()
+
+                    toml_cu_key = toml[cu_key]
+                    assert isinstance(
+                        toml_cu_key,
+                        (TOMLDocument, toml_items.Table, OutOfOrderTableProxy),
+                    ), f"key {cu_key}: attempting to recurse into an non-table/document"
+
+                    _apply_basemodel_differences(og_val, cu_val, toml_cu_key)
+
+                elif og_val != cu_val:
+                    toml[cu_key] = cu_val  # pyright: ignore[reportAny] (from pydantic)
+
         # add values that were changed from when the model was instantiated
         def apply_model_differences(
             original_model: BaseModel,
@@ -683,15 +748,78 @@ class ModelBoundTOML(Generic[M]):
                         toml[cu_key] = toml_table()
 
                     toml_cu_key = toml[cu_key]
-                    assert isinstance(toml_cu_key, (TOMLDocument, toml_items.Table)), (
-                        f"key {cu_key}: attempting to recurse into an non-table/document"
-                    )
+                    assert isinstance(
+                        toml_cu_key,
+                        (TOMLDocument, toml_items.Table, OutOfOrderTableProxy),
+                    ), f"key {cu_key}: attempting to recurse into an non-table/document"
 
                     apply_model_differences(og_val, cu_val, toml_cu_key)
 
                 elif original_kv != current_kv:
                     cu_key, cu_val = current_kv  # pyright: ignore[reportAny] (from pydantic)
-                    toml[cu_key] = cu_val
+
+                    if _is_dict_of_basemodels(cu_val) and _is_dict_of_basemodels(
+                        og_val
+                    ):
+                        if _dicts_of_basemodels_differ(og_val, cu_val):  # type: ignore[arg-type]
+                            if cu_key not in toml:
+                                toml[cu_key] = toml_table()
+
+                            toml_cu_key = toml[cu_key]
+                            assert isinstance(
+                                toml_cu_key, (TOMLDocument, toml_items.Table)
+                            ), (
+                                f"key {cu_key}: attempting to set table of tables in non-table/document"
+                            )
+
+                            cu_dict = cu_val  # type: ignore[assignment]
+                            og_dict = og_val  # type: ignore[assignment]
+                            for dict_key, dict_val in cu_dict.items():
+                                if (
+                                    dict_key not in og_dict
+                                    or og_dict[dict_key] != dict_val
+                                ):
+                                    if dict_key not in toml_cu_key:
+                                        toml_cu_key[dict_key] = toml_table()
+
+                                    toml_dict_table = toml_cu_key[dict_key]
+                                    assert isinstance(
+                                        toml_dict_table,
+                                        (TOMLDocument, toml_items.Table),
+                                    ), (
+                                        f"key {cu_key}.{dict_key}: attempting to set non-table/document"
+                                    )
+
+                                    if dict_key in og_dict:
+                                        _apply_basemodel_differences(
+                                            original_model=og_dict[dict_key],
+                                            current_model=dict_val,
+                                            toml=toml_dict_table,
+                                        )
+                                    else:
+                                        for val_key, val_val in dict_val:
+                                            toml_dict_table[val_key] = val_val  # pyright: ignore[reportAny]
+
+                    elif _is_list_of_basemodels(cu_val) and _is_list_of_basemodels(
+                        og_val
+                    ):
+                        if _lists_of_basemodels_differ(og_val, cu_val):  # type: ignore[arg-type]
+                            if cu_key not in toml:
+                                toml[cu_key] = toml_array()
+
+                            toml_array_val = toml[cu_key]
+                            assert isinstance(
+                                toml_array_val, (toml_items.Array, toml_items_AoT)
+                            ), f"key {cu_key}: value is not an array or array of tables"
+
+                            toml_array_val.clear()
+                            cu_list = cu_val  # type: ignore[assignment]
+                            for list_item in cu_list:
+                                item_dict = {k: v for k, v in list_item}  # pyright: ignore[reportAny]
+                                toml_array_val.append(item_dict)
+
+                    else:
+                        toml[cu_key] = cu_val
 
         apply_model_differences(self.__original_model, self.model, document)
         return document
@@ -817,7 +945,10 @@ class ModelBoundTOML(Generic[M]):
 
                     incoming_value = incoming_document[outgoing_key]
 
-                    if not isinstance(incoming_value, (TOMLDocument, toml_items.Table)):
+                    if not isinstance(
+                        incoming_value,
+                        (TOMLDocument, toml_items.Table, OutOfOrderTableProxy),
+                    ):
                         # if the model is a BaseModel but the document is not a table,
                         # the difference is incoming because we assume the model is the
                         # source of truth

